In [1]:
%run 10_ngram_common.ipynb

/work/van-speech-nlp/jindaznb/asrenv/lib/python3.10/site-packages/huggingface_hub/utils/_runtime.py:184: UserWarning: Pydantic is installed but cannot be imported. Please check your installation. `huggingface_hub` will default to not using Pydantic. Error message: '{e}'
  warnings.warn(


In [2]:
normal_speakers = ['MC01', 'MC02', 'MC03', 'MC04','FC01','FC02','FC03']
atypical_speakers = ['F01', 'F03', 'F04', 'M01', 'M02', 'M03', 'M04', 'M05']

# read europarl

In [3]:
europarl = load_dataset(f"{username}/{source_lang}_corpora_parliament_processed", split="train",cache_dir="/work/van-speech-nlp/cache")
translations=europarl['text']

Generating train split:   0%|          | 0/2051014 [00:00<?, ? examples/s]

In [14]:
len(translations)

2051014

In [15]:
import nltk
total_tokens = 0

# Iterate over each sentence in the translations corpus
for sentence in translations:
    # Tokenize the sentence into individual words
    tokens = nltk.word_tokenize(sentence)
    # Add the number of tokens in the current sentence to the total count
    total_tokens += len(tokens)

# Print the total number of tokens in the translations corpus
print("Total number of tokens in corpus:", total_tokens)

Total number of tokens in corpus: 51953865


In [4]:
word_set_europarl = set()

for sentence in translations:
    words = sentence.split()
    word_set_europarl.update(words)

In [6]:
len(word_set_europarl)

120001

In [7]:
filtered_word_set_cleaned = set()

for word in translations:
    if word:
        cleaned_word = re.sub(r'[^a-zA-Z0-9]', '', word.lower())
        filtered_word_set_cleaned.add(cleaned_word)

cleaned_word_count = len(filtered_word_set_cleaned)

In [8]:
cleaned_word_count

1994288

# read torgo

In [13]:
pattern = "keep_all"

In [8]:
# Read the dataset
data_df = pd.read_csv('torgo.csv')
dataset_csv = load_dataset('csv', data_files='torgo.csv')

speakers = data_df['speaker_id'].unique()

print(f'Speakers: {", ".join(speakers)}')

Generating train split: 0 examples [00:00, ? examples/s]

Speakers: F01, F03, F04, FC01, FC02, FC03, M01, M02, M03, M04, M05, MC01, MC02, MC03, MC04


In [14]:
dataset_csv

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

Filter:   0%|          | 0/16394 [00:00<?, ? examples/s]

F01: 15091
F03: 14652
F04: 14652
M01: 14580
M02: 14553
M03: 14519
M04: 14667
M05: 14746


# calculate unique words

In [16]:
unique_words_per_speaker = {}

for speaker_id, texts in speaker_texts.items():
    unique_words = set()
    for text in texts:
        words = text.split()
        for word in words:
            cleaned_word = re.sub(r'[^a-zA-Z0-9]', '', word.lower())
            if cleaned_word:  
                unique_words.add(cleaned_word) 
    unique_words_per_speaker[speaker_id] = unique_words

for speaker_id, unique_words in unique_words_per_speaker.items():
    print(f"{speaker_id}: {len(unique_words)} unique words")

F01: 1561 unique words
F03: 1557 unique words
F04: 1557 unique words
M01: 1559 unique words
M02: 1557 unique words
M03: 1561 unique words
M04: 1559 unique words
M05: 1557 unique words


# vocabulary similarity

In [17]:
out_of_domain_words_per_speaker = {}

for speaker_id, unique_words in unique_words_per_speaker.items():
    overlap = len(unique_words & word_set_europarl)
    similarity = overlap / len(unique_words) if len(unique_words) > 0 else 0
    out_of_domain_similarity = (1 - similarity)*100

    out_of_domain_words = {word for word in unique_words if word not in word_set_europarl}
    out_of_domain_words_per_speaker[speaker_id] = out_of_domain_words

    print(f"{speaker_id}: {out_of_domain_similarity:.2f}% dissimilarity with word_set_europarl")
    print(f"{speaker_id}: out_of_domain_words count: {len(out_of_domain_words)}")
    print(f"Out of domain words for {speaker_id}: {out_of_domain_words}")
    print()

F01: 5.70% dissimilarity with word_set_europarl
F01: out_of_domain_words count: 89
Out of domain words for F01: {'dill', 'scrubbed', 'marvelously', 'ambidextrous', 'sitter', 'wack', 'werent', 'weve', 'owl', 'bongos', 'quivers', 'briar', 'touchdown', 'taxaided', 'lacy', 'gingham', 'postdate', 'musk', 'screech', 'woolen', 'youd', 'trespassers', 'ligament', 'porch', 'toot', 'nancys', 'bender', 'innocense', 'thigh', 'jims', 'shawl', 'whoop', 'calico', 'advisement', 'mall', 'slicker', 'combed', 'mabel', 'giggling', 'kat', 'solicits', 'glitter', 'aluminum', 'frock', 'accordion', 'rakes', 'arent', 'ooze', 'darn', 'yell', 'treehouse', 'droop', 'sprain', 'schoolhouse', 'bloat', 'mold', 'urchins', 'ail', 'xray', 'hem', 'fleecy', 'boone', 'renter', 'chowder', 'lilyrare', 'stubble', 'someplace', 'whitecapped', 'youre', 'mom', 'beep', 'bracelet', 'colorful', 'dart', 'porcupines', 'theyll', 'ladle', 'prowler', 'gash', 'snort', 'appetizers', 'pete', 'fender', 'coupe', 'bun', 'ive', 'mut', 'shimmers',

# Out of domain analysis

In [47]:
import kenlm

model=kenlm.Model("kenlm_model_3gram_words_europarl/3gram.bin") 
per=model.perplexity("your text sentance")
print(per)

per=model.perplexity("resumption of the session")
print(per)

2445.298548730424
13.064232719071327


In [54]:
# Calculate perplexity for each text
perplexities = []
for text in cleaned_texts:
    perplexity = model.perplexity(text)
    perplexities.append(perplexity)

# Calculate the average perplexity
avg_perplexity = np.mean(perplexities)
print(f"Average Perplexity: {avg_perplexity}")

Average Perplexity: 1498.1788760387824


# Average Perplexity: 1498.1788760387824 showing that TORGO is out of domain text

# clean the text

In [58]:
import re

cleaned_speaker_texts = {}

for speaker_id, texts in speaker_texts.items():
    cleaned_texts = []
    for text in texts:
        cleaned_text = ' '.join(re.sub(r'[^a-zA-Z0-9]', '', word.lower()) for word in text.split())
        cleaned_texts.append(cleaned_text)
    cleaned_speaker_texts[speaker_id] = cleaned_texts

for speaker_id, texts in cleaned_speaker_texts.items():
    print(f"Speaker {speaker_id}:")
    for text in texts[:2]:  
        print(text)
    print()

Speaker F01:
stick
except in the winter when the ooze or snow or ice prevents

Speaker F03:
beta
stubble

Speaker F04:
fee
yet he still thinks as swiftly as ever

Speaker M01:
when he speaks his voice is just a bit cracked and quivers a trifle
trait

Speaker M02:
grow
feed

Speaker M03:
now
know

Speaker M04:
trouble
spark

Speaker M05:
swore
train



In [60]:
# Initialize a dictionary to store perplexities per speaker
perplexities_per_speaker = {}

# Calculate perplexity for each speaker
for speaker_id, texts in cleaned_speaker_texts.items():
    perplexities = []
    for text in texts:
        perplexity = model.perplexity(text)
        perplexities.append(perplexity)
    avg_perplexity = np.mean(perplexities)
    perplexities_per_speaker[speaker_id] = avg_perplexity

# Print average perplexity for each speaker
for speaker_id, avg_perplexity in perplexities_per_speaker.items():
    print(f"Speaker {speaker_id}: Average Perplexity: {avg_perplexity}")

Speaker F01: Average Perplexity: 1014.5739804044058
Speaker F03: Average Perplexity: 1468.7080678251675
Speaker F04: Average Perplexity: 1474.5388428766746
Speaker M01: Average Perplexity: 1534.7643705710645
Speaker M02: Average Perplexity: 1492.4557345179953
Speaker M03: Average Perplexity: 1492.9785709867167
Speaker M04: Average Perplexity: 1531.2270123869641
Speaker M05: Average Perplexity: 1584.9380105002526
